diagnostic/engine.go

```diff
@@ -12,18 +12,113 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package inference
+// Package diagnostic hosts the diagnostic engine, which is responsible for collecting the
+// conflicts from annotation-based checks (no-infer mode) and/or inference (full-infer mode) and
+// generating user-friendly diagnostics from those conflicts.
+package diagnostic
 
 import (
 	"fmt"
 	"go/token"
 	"strings"
 
 	"go.uber.org/nilaway/annotation"
+	"go.uber.org/nilaway/inference"
 	"go.uber.org/nilaway/util"
 	"golang.org/x/tools/go/analysis"
 )
 
+// Engine is the main engine for generating diagnostics from conflicts.
+type Engine struct {
+	pass      *analysis.Pass
+	conflicts []conflict
+}
+
+// NewEngine creates a new diagnostic engine.
+func NewEngine(pass *analysis.Pass) *Engine {
+	return &Engine{pass: pass}
+}
+
+// Diagnostics generates diagnostics from the internally-stored conflicts. The grouping parameter
+// controls whether the conflicts with the same nil path are grouped together for concise reporting.
+func (e *Engine) Diagnostics(grouping bool) []analysis.Diagnostic {
+	conflicts := e.conflicts
+	if grouping {
+		// group conflicts with the same nil path together for concise reporting
+		conflicts = groupConflicts(e.conflicts)
+	}
+
+	// build diagnostics from conflicts
+	diagnostics := make([]analysis.Diagnostic, 0, len(conflicts))
+	for _, c := range conflicts {
+		diagnostics = append(diagnostics, analysis.Diagnostic{
+			Pos:     c.pos,
+			Message: c.String(),
+		})
+	}
+	return diagnostics
+}
+
+// AddSingleAssertionConflict adds a new single assertion conflict to the engine.
+func (e *Engine) AddSingleAssertionConflict(trigger annotation.FullTrigger) {
+	producer, consumer := trigger.Prestrings(e.pass)
+	flow := nilFlow{}
```

Comment:
According to our [style guide](https://github.com/uber-go/guide/blob/master/style.md#function-grouping-and-ordering), we should group functions by receivers (i.e., methods should follow the the type declarations), where the exported types and functions should appear earlier than unexported ones.

Actually, since we are now separating the diagnostic logic into a new package, we can just put the `conflict` and `nilFlow` type declarations into their own files for easier navigation.

Hopefully now it's clearer and easier to find them 😃